packages/xstate-react/src/useMachine.ts

```diff
@@ -81,21 +76,36 @@ export function useMachine<
         nextState.changed === undefined &&
         Object.keys(nextState.children).length;
 
-      if (nextState.changed || initialStateChanged) {
-        setState(nextState);
-      }
+      return !(nextState.changed || initialStateChanged);
+    },
+    [service]
+  );
+
+  const subscribe = useCallback(
+    (handleStoreChange) => {
+      const { unsubscribe } = service.subscribe(handleStoreChange);
+      return unsubscribe;
     },
-    []
+    [service]
+  );
+  const storeSnapshot = useSyncExternalStoreWithSelector(
```

Comment:
`useMachine` kinda shouldn't have to use `useSyncExternalStoreWithSelector` - it could use the simpler `useSyncExternalStore`, but to make this possible we'd have to have immutable states without `.changed`